File: python/elements/replicable_graph_element.py
from elements import ReplicaState, GraphElement, GraphQuery, Op, query_for_part
from typing import TYPE_CHECKING, Tuple
from copy import copy
from elements import ElementTypes

if TYPE_CHECKING:
    from elements.element_feature.set_feature import SetReplicatedFeature
    from elements.rpc import Rpc


class ReplicableGraphElement(GraphElement):
    """ General Graph element that is replicatable. Used in Vertex for now can be added for other things as well """

    def __init__(self, element_id: str = None, master: int = None, is_halo=False, *args, **kwargs):
        super(ReplicableGraphElement, self).__init__(element_id=element_id, *args, **kwargs)
        self._master = master  # Part_id of the master
        self._halo = is_halo  # If this only a stub for master

    def __call__(self, rpc: "Rpc") -> Tuple[bool, "GraphElement"]:
        """ Wrap GraphElement call to have separate behavior for Replica & Master nodes """
        is_updated, elem = super(ReplicableGraphElement, self).__call__(rpc)
        if is_updated:
            self.sync_replicas()
        return is_updated, elem

    def create_element(self) -> bool:
        if self.state is ReplicaState.REPLICA: self._features.clear()  # Clear needed since it will be synced with
        # master anyway
        is_created = super(ReplicableGraphElement, self).create_element()  # Store addition and callbacks
        if not is_created: return is_created  # Failed to create or already existed in the database
        if self.state is ReplicaState.MASTER:
            from elements.element_feature.set_feature import SetReplicatedFeature
            self['parts'] = SetReplicatedFeature({self.storage.part_id}, is_halo=True)  # No need to replicate
            # replicas do not need this
        elif self.state is ReplicaState.REPLICA:
            # Sync this element with its master
            query = GraphQuery(Op.SYNC, self, self.master_part, True)
            self.storage.message(query)
        return is_created

    def sync_element(self, new_element: "GraphElement") -> Tuple[bool, "GraphElement"]:
        if self.state is ReplicaState.MASTER:
            # Add to parts of replicas and sync with this part
            self["parts"].add(new_element.part_id)
            self.sync_replicas(new_element.part_id, skip_halos=False)
            return False, self
        elif self.state is ReplicaState.REPLICA:
            # Commit the update to replica
            return self.update_element(new_element)

    def external_update(self, new_element: "GraphElement") -> Tuple[bool, "GraphElement"]:
        if self.state is ReplicaState.REPLICA:
            # External Updates only done in master part
            query = GraphQuery(Op.ADDUPDATE, new_element, self.master_part, True)
            self.storage.message(query)
            return False, self
        elif self.state is ReplicaState.MASTER:
            is_updated, memento = super(ReplicableGraphElement, self).external_update(
                new_element)  # Basically calling update_element
            if is_updated:
                self.sync_replicas()
            return is_updated, memento

    def __iter__(self):
        """ Do not have parts in the iter  """
        return super(ReplicableGraphElement, self).__iter__()

    def sync_replicas(self, part_id=None, skip_halos=True):
        """ Sending this element to all or some replicas. This element is shallow copied for operability """
        if self.state is not ReplicaState.MASTER or (skip_halos and self.is_halo) or (len(
                self.replica_parts) == 0 and not part_id): return
        self.cache_features()
        cpy_self = copy(self)
        cpy_self._features.clear()
        for key, value in self:
            if skip_halos and value.is_halo:
                continue
            ft_cpy = copy(value)
            ft_cpy.element = cpy_self
            cpy_self._features[key] = ft_cpy
            if ft_cpy.is_halo: ft_cpy._value = None

        query = GraphQuery(op=Op.SYNC, element=cpy_self, part=None, iterate=True)

        if part_id:
            self.storage.message(query_for_part(query, part_id))
        else:
            filtered_parts = map(lambda x: query_for_part(query, x), self.replica_parts)
            for msg in filtered_parts:
                self.storage.message(msg)

    @property
    def master_part(self) -> int:
        return self._master

    @property
    def is_halo(self) -> bool:
        return self._halo

    @property
    def replica_parts(self) -> list:
        try:
            re: "SetReplicatedFeature" = self['parts']
            a = list(re.value)
            a.remove(self.part_id)
            return a
        except KeyError:
            return list()

    @property
    def is_replicable(self) -> bool:
        return True

    def __copy__(self):
        element = super(ReplicableGraphElement, self).__copy__()
        element.__dict__.update({
            "_master": self.master_part,
            "_halo": self.is_halo
        })
        return element

    def __deepcopy__(self, memodict={}):
        element = super(ReplicableGraphElement, self).__deepcopy__()
        element.__dict__.update({
            "_master": self.master_part,
            "_halo": self.is_halo
        })
        return element

    def __getstate__(self):
        """ No need to serialize the parts """
        state = super(ReplicableGraphElement, self).__getstate__()
        state.update({
            "_master": self.master_part,
            "_halo": self.is_halo
        })
        return state

    def __get_save_data__(self):
        metadata = super(ReplicableGraphElement, self).__get_save_data__()
        metadata.update({
            "_master": self.master_part,
            "_halo": self.is_halo
        })
        return metadata

File: src/lib/java/elements/ReplicableGraphElement.java
package elements;

import elements.enums.*;
import elements.features.Parts;
import it.unimi.dsi.fastutil.shorts.ShortArrayList;
import org.apache.flink.streaming.api.operators.graph.OutputTags;

import java.util.List;

/**
 * {@link GraphElement} that are replicable. Added master sync logic
 *
 * @implNote {@link Feature} is a {@link ReplicableGraphElement}
 * but if it is attached to non-replicable {@link GraphElement} it will behave so as well
 */
abstract public class ReplicableGraphElement extends GraphElement {

    /**
     * Master part of this element
     */
    public short masterPart = -1;

    public ReplicableGraphElement() {
        super();
    }

    public ReplicableGraphElement(short masterPart) {
        super();
        this.masterPart = masterPart;
    }

    public ReplicableGraphElement(ReplicableGraphElement element, CopyContext copyContext) {
        super(element, copyContext);
        this.masterPart = element.masterPart;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    abstract public ReplicableGraphElement copy(CopyContext context);

    /**
     * {@inheritDoc}
     * <p>
     * If this is REPLICA clearFeatures & if not halo --> send {@link SyncRequest} to Master before running callbacks
     * Since we cannot always assume that Master exists (because of possible lateness),
     * the {@link SyncRequest} should arrive before any other output is triggered in {@link Plugin} to perform dummy element creation
     * </p>
     */
    @Override
    public void create() {
        if (state() == ReplicaState.REPLICA && features != null) features.clear();
        if (!isHalo() && isReplicable() && state() == ReplicaState.REPLICA) {
            GraphOp syncRequestMessage = new GraphOp(Op.SYNC_REQUEST, getMasterPart(), new SyncRequest(this));
            getGraphRuntimeContext().output(syncRequestMessage, OutputTags.ITERATE_OUTPUT_TAG);
        }
        super.create();
    }

    /**
     * {@inheritDoc}
     * Update without replication
     */
    @Override
    public void sync(GraphElement newElement) {
        updateInternal(newElement);
    }

    /**
     * {@inheritDoc}
     * <p>
     * Add newly arrived part number to the <strong>parts</strong> list
     * Only sends back a SYNC if there are non-halo {@link Feature} or if this is a STANDALONE {@link Feature} itself
     * </p>
     */
    @Override
    public void syncRequest(GraphElement newElement) {
        if (!containsFeature("p")) {
            Parts p = new Parts("p", new ShortArrayList(List.of(newElement.getPart())), true);
            p.setElement(this, false);
            p.createInternal();
        } else {
            Rmi.execute(
                    getFeature("p"),
                    "add",
                    newElement.getPart()
            );
        }
        if ((features != null && features.stream().anyMatch(feature -> !feature.isHalo())) || getType() == ElementType.STANDALONE_FEATURE) {
            GraphElement cpy = copy(CopyContext.SYNC);
            getGraphRuntimeContext().output(new GraphOp(Op.SYNC, newElement.getPart(), cpy), OutputTags.ITERATE_OUTPUT_TAG);
        }
    }

    /**
     * {@inheritDoc}
     *
     * @implNote REPLICA element receiving updates will trigger {@link IllegalStateException} since not allowed
     * <p>
     * If this newElement is a Feature or brings some non-halo features with it, before executing update will send SYNC to all replicas
     * </p>
     */
    @Override
    public void update(GraphElement newElement) {
        if (state() == ReplicaState.MASTER) {
            if (!isHalo() && isReplicable() && !getReplicaParts().isEmpty() && (getType() == ElementType.ATTACHED_FEATURE || getType() == ElementType.STANDALONE_FEATURE || (newElement.features != null && newElement.features.stream().anyMatch(feature -> !feature.isHalo())))) {
                GraphOp message = new GraphOp(Op.SYNC, newElement.copy(CopyContext.SYNC));
                getGraphRuntimeContext().broadcast(message, OutputTags.ITERATE_OUTPUT_TAG, getReplicaParts());
            }
            super.update(newElement);
        } else {
            throw new IllegalStateException("REPLICAS Should not received Updates");
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public short getMasterPart() {
        return this.masterPart;
    }

    /**
     * {@inheritDoc}
     */
    @SuppressWarnings("unchecked")
    @Override
    public List<Short> getReplicaParts() {
        if (!containsFeature("p")) return super.getReplicaParts();
        return (List<Short>) (getFeature("p")).getValue(); // @implNote Never create other Feature with the name parts
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public String toString() {
        return getType() + "{" +
                "id='" + getId() + '\'' +
                "master='" + masterPart + '\'' +
                '}';
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public boolean isReplicable() {
        return true;
    }
}
